File: app/utils/timezone.py
"""
Timezone utilities for BayScan.

Ensures all times displayed to users are in Central Standard Time (America/Chicago).
All internal storage remains in UTC.
"""
from datetime import datetime
from zoneinfo import ZoneInfo

# Central timezone
CENTRAL_TZ = ZoneInfo("America/Chicago")
UTC_TZ = ZoneInfo("UTC")


def utc_to_central(utc_dt: datetime) -> datetime:
    """
    Convert UTC datetime to Central time.

    Args:
        utc_dt: Naive or UTC datetime object

    Returns:
        Datetime in Central timezone (America/Chicago)
    """
    if utc_dt is None:
        return None

    # If naive, assume UTC
    if utc_dt.tzinfo is None:
        utc_dt = utc_dt.replace(tzinfo=UTC_TZ)

    # Convert to Central
    return utc_dt.astimezone(CENTRAL_TZ)


def format_central_time(utc_dt: datetime, include_timezone: bool = True) -> str:
    """
    Format UTC datetime as Central time string.

    Args:
        utc_dt: UTC datetime object
        include_timezone: Whether to append " CST" or " CDT" to the string

    Returns:
        Formatted string like "3:45 PM CST" or "3:45 PM"
    """
    if utc_dt is None:
        return ""

    central_dt = utc_to_central(utc_dt)
    time_str = central_dt.strftime("%I:%M %p").lstrip("0")  # Remove leading zero from hour

    if include_timezone:
        # Get timezone abbreviation (CST or CDT)
        tz_abbr = central_dt.strftime("%Z")
        return f"{time_str} {tz_abbr}"

    return time_str


def format_central_datetime(utc_dt: datetime, include_timezone: bool = True) -> str:
    """
    Format UTC datetime as Central date and time string.

    Args:
        utc_dt: UTC datetime object
        include_timezone: Whether to append timezone abbreviation

    Returns:
        Formatted string like "Nov 23, 3:45 PM CST"
    """
    if utc_dt is None:
        return ""

    central_dt = utc_to_central(utc_dt)
    date_str = central_dt.strftime("%b %d")
    time_str = central_dt.strftime("%I:%M %p").lstrip("0")

    if include_timezone:
        tz_abbr = central_dt.strftime("%Z")
        return f"{date_str}, {time_str} {tz_abbr}"

    return f"{date_str}, {time_str}"


def get_central_isoformat(utc_dt: datetime) -> str:
    """
    Convert UTC datetime to ISO format string with Central timezone.

    Args:
        utc_dt: UTC datetime object

    Returns:
        ISO format string with timezone offset (e.g., "2024-11-23T14:30:00-06:00")
    """
    if utc_dt is None:
        return ""

    central_dt = utc_to_central(utc_dt)
    return central_dt.isoformat()


def central_now() -> datetime:
    """
    Get current datetime in Central timezone.

    Returns:
        Current datetime in America/Chicago timezone
    """
    return datetime.now(CENTRAL_TZ)


def central_to_utc(central_dt: datetime) -> datetime:
    """
    Convert Central datetime to UTC for database storage.

    Args:
        central_dt: Naive datetime representing Central time, or aware Central datetime

    Returns:
        UTC datetime (naive, for database storage)
    """
    if central_dt is None:
        return None

    # If naive, treat as Central time
    if central_dt.tzinfo is None:
        central_dt = central_dt.replace(tzinfo=CENTRAL_TZ)

    # Convert to UTC and make naive for database storage
    utc_dt = central_dt.astimezone(UTC_TZ)
    return utc_dt.replace(tzinfo=None)

File: app/static/js/timezone.js
/**
 * Timezone utilities for BayScan frontend
 * Ensures all times display in Central Time (America/Chicago)
 */

/**
 * Format a datetime string (ISO format) to Central time
 * @param {string} isoString - ISO format datetime string with timezone
 * @param {object} options - Formatting options
 * @returns {string} Formatted time string
 */
function formatCentralTime(isoString, options = {}) {
    if (!isoString) return '';

    const date = new Date(isoString);
    const defaultOptions = {
        hour: 'numeric',
        minute: '2-digit',
        hour12: true,
        timeZone: 'America/Chicago'
    };

    const formatOptions = { ...defaultOptions, ...options };
    return date.toLocaleTimeString('en-US', formatOptions);
}

/**
 * Format a datetime string to Central time with timezone abbreviation
 * @param {string} isoString - ISO format datetime string
 * @param {boolean} showTimezone - Whether to show CST/CDT abbreviation
 * @returns {string} Formatted time like "3:45 PM CST"
 */
function formatCentralTimeWithZone(isoString, showTimezone = true) {
    if (!isoString) return '';

    const date = new Date(isoString);
    const timeStr = date.toLocaleTimeString('en-US', {
        hour: 'numeric',
        minute: '2-digit',
        hour12: true,
        timeZone: 'America/Chicago'
    });

    if (!showTimezone) {
        return timeStr;
    }

    // Get timezone abbreviation (CST or CDT)
    const tzStr = date.toLocaleTimeString('en-US', {
        timeZone: 'America/Chicago',
        timeZoneName: 'short'
    });

    // Extract timezone abbreviation from string like "3:45:00 PM CST"
    const tzMatch = tzStr.match(/\s([A-Z]{3})$/);
    const tzAbbr = tzMatch ? tzMatch[1] : 'CST';

    return `${timeStr} ${tzAbbr}`;
}

/**
 * Format a datetime string to Central date
 * @param {string} isoString - ISO format datetime string
 * @param {object} options - Formatting options
 * @returns {string} Formatted date string
 */
function formatCentralDate(isoString, options = {}) {
    if (!isoString) return '';

    const date = new Date(isoString);
    const defaultOptions = {
        month: 'short',
        day: 'numeric',
        timeZone: 'America/Chicago'
    };

    const formatOptions = { ...defaultOptions, ...options };
    return date.toLocaleDateString('en-US', formatOptions);
}

/**
 * Format a datetime string to Central date and time
 * @param {string} isoString - ISO format datetime string
 * @param {boolean} showTimezone - Whether to show timezone abbreviation
 * @returns {string} Formatted string like "Nov 23, 3:45 PM CST"
 */
function formatCentralDateTime(isoString, showTimezone = true) {
    if (!isoString) return '';

    const dateStr = formatCentralDate(isoString);
    const timeStr = formatCentralTimeWithZone(isoString, showTimezone);

    return `${dateStr}, ${timeStr}`;
}

/**
 * Get current time in Central timezone
 * @returns {Date} Current date/time
 */
function getCentralNow() {
    return new Date(new Date().toLocaleString('en-US', { timeZone: 'America/Chicago' }));
}

// Backward compatibility: Update the existing formatTime function
// This will be used throughout the existing codebase
function formatTime(dateOrString) {
    // Handle both Date objects and ISO strings
    if (typeof dateOrString === 'string') {
        return formatCentralTimeWithZone(dateOrString, false);
    } else if (dateOrString instanceof Date) {
        return dateOrString.toLocaleTimeString('en-US', {
            hour: 'numeric',
            minute: '2-digit',
            hour12: true,
            timeZone: 'America/Chicago'
        });
    }
    return '';
}
